simSPI/linear_simulator/projector.py

```diff
@@ -21,18 +21,144 @@ def __init__(self, config):
         super(Projector, self).__init__()
 
         self.config = config
-        self.vol = torch.rand([self.config.side_len] * 3, dtype=torch.float32)
-        lin_coords = torch.linspace(-1.0, 1.0, self.config.side_len)
-        [x, y, z] = torch.meshgrid(
-            [
-                lin_coords,
-            ]
-            * 3
-        )
-        coords = torch.stack([y, x, z], dim=-1)
-        self.register_buffer("vol_coords", coords.reshape(-1, 3))
+        self.space = config.space
+
+        if self.space == "real":
+            self.vol = torch.rand([self.config.side_len] * 3, dtype=torch.float32)
+            lin_coords = torch.linspace(-1.0, 1.0, self.config.side_len)
+            [x, y, z] = torch.meshgrid(
+                [
+                    lin_coords,
+                ]
+                * 3
+            )
+            coords = torch.stack([y, x, z], dim=-1)
+
+            self.register_buffer("vol_coords", coords.reshape(-1, 3))
+        elif self.space == "fourier":
+            # Assume DC coefficient is at self.vol[n//2+1,n//2+1]
+            # this means that self.vol = fftshift(fft3(fftshift(real_vol)))
+            self.vol = torch.rand([self.config.side_len] * 3, dtype=torch.complex64)
+            freq_coords = torch.fft.fftfreq(self.config.side_len, dtype=torch.float32)
+            [x, y] = torch.meshgrid(
+                [
+                    freq_coords,
+                ]
+                * 2
+            )
+            coords = torch.stack([y, x], dim=-1)
+            # Rescale coordinates to [-1,1] to be compatible with
+            # torch.nn.functional.grid_sample
+            coords = 2 * coords
+            self.register_buffer("vol_coords", coords.reshape(-1, 2))
+        else:
+            raise NotImplementedError(
+                f"Space type '{self.space}' " f"has not been implemented!"
+            )
 
     def forward(self, rot_params, proj_axis=-1):
+        """Forward method for projection.
+
+        Parameters
+        ----------
+        rot_params : tensor of rotation matrices
```

Comment:
The description of `rot_params` does not strictly follow the docstring convention:
- if it is a tensor, what is its shape
- it should be on two lines, where the first line explains the type of datastructure and the second explains what the parameter represents.

This will become important to generate a clean documentation website.

Additionally, this description says that rot_params is a tensor; yet `_forward_fourier` docstring mentions a dict: which one is true?

Reply:
Maybe this docstring could point to `_forward_fourier` docstring about the details? I find it well explained there that `rot_params` is a dictionary that contains a tensor of specific shape.